### PR TITLE
PowerService: UartFraming/CobsEncoding

### PR DESCRIPTION
**Pull request by LiCody**
### Description

- Implemented COBS encoding and decoding algorithm
- Added crc16 and length framing for uart messages

### Testing Done

- added unit tests

### Resolved Issues
Part of #2334

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/shared/uart_framing/uart_framing.hpp
Comment (akhilveeraghanta) [inline src/shared/uart_framing/uart_framing.hpp]: curious why we template here and then cast to an unsigned byte? why not just take an uint8_t pointer to start and leave it up to the user to cast if needed

File: src/shared/uart_framing/uart_framing.hpp
Comment (akhilveeraghanta) [inline src/shared/uart_framing/uart_framing.hpp]: this is neat!

File: src/shared/uart_framing/uart_framing.hpp
Comment (akhilveeraghanta) [inline src/shared/uart_framing/uart_framing.hpp]: This is creating a new message so maybe just rename this to `createUartMessageFrame`, itll be a UartMessage Frame "factory"

File: src/shared/uart_framing/uart_framing.hpp
Comment (akhilveeraghanta) [inline src/shared/uart_framing/uart_framing.hpp]: thoughts on moving marshall and unmarshall to be functions of UartMessageFrame?

File: src/shared/uart_framing/uart_framing_test.cpp
Comment (akhilveeraghanta) [inline src/shared/uart_framing/uart_framing_test.cpp:52]: nice tests!

File: src/shared/uart_framing/uart_framing_test.cpp
Comment (akhilveeraghanta) [inline src/shared/uart_framing/uart_framing_test.cpp:91]: do we need to add `__attribute__((__packed__))__` for this to work? have you tried protobuf?

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
awesome work! really love the tests, and the implementation is very clean :100: 
this will be the glue between our main computer and the power board for years to come

just had a couple comments

File: src/shared/uart_framing/uart_framing.hpp
Comment (amrm3lm) [inline src/shared/uart_framing/uart_framing.hpp]: can you clarify what you mean by different/normal?

File: src/shared/uart_framing/uart_framing.hpp
Comment (amrm3lm) [inline src/shared/uart_framing/uart_framing.hpp]: curious what the idea is behind the upper bound calculation

File: src/shared/uart_framing/uart_framing_test.cpp
Comment (amrm3lm) [inline src/shared/uart_framing/uart_framing_test.cpp:67]: can you add a test which checks the case of the vector being larger than 0xFF, and also a test where there is a START_END_FLAG_BYTE between normal values

File: src/shared/uart_framing/uart_framing_test.cpp
Comment (amrm3lm) [reply src/shared/uart_framing/uart_framing_test.cpp:91]: Also would this lead to complications given that the Nano uses a different architecture? if `__attribute__((__packed__))__` is needed then i would consider using a byte array, which we can get from protobuf

**Review by amrm3lm — COMMENTED** [review]
really cool stuff!! just wanted clarification on some things, and thought of a couple more cases worth testing

File: src/shared/uart_framing/uart_framing.hpp
Comment (jonathanlew) [inline src/shared/uart_framing/uart_framing.hpp]: what do of 3 and 254 represent?

**Review by jonathanlew — COMMENTED** [review]
not much to add beyond existing comments

File: src/shared/uart_framing/uart_framing.hpp
Comment (LiCody) [reply src/shared/uart_framing/uart_framing.hpp]: Moving marshal to be a function of UartMessageFrame makes sense to me but unmarshalling doesn't because it returns an optional.

File: src/shared/uart_framing/uart_framing_test.cpp
Comment (LiCody) [reply src/shared/uart_framing/uart_framing_test.cpp:91]: adding __attribute__((__packed__))__ makes the messages smaller sometimes but we should be fine to remove it if we do encounter issues with it.

File: src/shared/uart_framing/uart_framing.hpp
Comment (LiCody) [reply src/shared/uart_framing/uart_framing.hpp]: different/normal when compared to the COBS algorithm on the wiki page. Adding text to clarify

File: src/shared/uart_framing/uart_framing.hpp
Comment (LiCody) [reply src/shared/uart_framing/uart_framing.hpp]: On second thought this check is not necessary. idea is 2 bytes at start and end for flag, 1 byte for flag and maximum n/254 additionally added bytes

File: src/shared/uart_framing/uart_framing_test.cpp
Comment (LiCody) [reply src/shared/uart_framing/uart_framing_test.cpp:67]: this behaviour/case actually happens when vectors are larger than 254/0xFE. added the other test case

**Review by jonathanlew — APPROVED** [review]
LGTM pending other reviewers' comments

**Review by akhilveeraghanta — APPROVED** [review]
Lgtm! Nice work 💯

**Review by amrm3lm — APPROVED** [review]
lgtm!

**Pull request merged**
